tools/bug-report/pkg/bugreport/bugreport.go

```diff
@@ -0,0 +1,253 @@
+			continue
+		}
+		writeFile(archive.ProxyLogPath(tempDir, namespace, pod), text)
+	}
+	return nil
+}
+
+// gatherInfo fetches all logs, resources, debug etc. using goroutines.
+// proxy logs and info are saved in logs/stats/importance global maps.
+// Errors are reported through gErrors.
+func gatherInfo(config *config.BugReportConfig, resources *cluster2.Resources, paths []string) {
+	var wg sync.WaitGroup
+	cmdTimer := time.NewTimer(time.Duration(config.CommandTimeout))
+
+	getFromCluster(content.GetK8sResources, &content.Params{DryRun: config.DryRun}, archive.ClusterInfoPath(tempDir, "resources"), &wg)
+	getFromCluster(content.GetCRs, &content.Params{DryRun: config.DryRun}, archive.ClusterInfoPath(tempDir, "crs"), &wg)
+	getFromCluster(content.GetEvents, &content.Params{DryRun: config.DryRun}, archive.ClusterInfoPath(tempDir, "events"), &wg)
+	getFromCluster(content.GetClusterInfo, &content.Params{DryRun: config.DryRun}, archive.ClusterInfoPath(tempDir, "cluster-info"), &wg)
+	getFromCluster(content.GetSecrets, &content.Params{DryRun: config.DryRun, Verbose: config.FullSecrets}, archive.ClusterInfoPath(tempDir, "secrets"), &wg)
+	getFromCluster(content.GetDescribePods, &content.Params{DryRun: config.DryRun, Namespace: config.IstioNamespace}, archive.ClusterInfoPath(tempDir, "describe-pods"), &wg)
+
+	var wg2 sync.WaitGroup
+	for _, p := range paths {
+		p := p
+		namespace, _, pod, container, err := cluster2.ParsePath(p)
+		if err != nil {
+			log.Error(err.Error())
+			continue
+		}
+		wg2.Add(1)
+		go func() {
+			defer wg2.Done()
+			var wg3 sync.WaitGroup
+
+			switch {
+			case container == "istio-proxy":
+				getFromCluster(content.GetCoredumps, &content.Params{DryRun: config.DryRun, Namespace: namespace, Pod: pod, Container: container},
+					archive.ProxyCoredumpPath(tempDir, namespace, pod), &wg3)
+				getProxyLogs(config, resources, p, namespace, pod, container, &wg3)
+
+			case strings.HasPrefix(pod, "istiod-") && container == "discovery":
+				getFromCluster(content.GetIstiodInfo, &content.Params{DryRun: config.DryRun, Namespace: namespace, Pod: pod, Container: container},
+					archive.IstiodPath(tempDir, namespace, pod+".debug"), &wg3)
+				getIstiodLogs(config, resources, namespace, pod, &wg3)
+
+			}
+			wg3.Wait()
+		}()
+	}
+
+	// Not all items are subject to timeout. Proceed only if the non-cancellable items have completed.
+	wg.Wait()
+
+	// If log fetches have completed, cancel the timeout.
+	go func() {
+		wg2.Wait()
+		cmdTimer.Stop()
+	}()
+
+	// Wait for log fetches, up to the timeout.
+	<-cmdTimer.C
+}
+
+// getFromCluster runs a cluster info fetching function f against the cluster and writes the results to fileName.
+// Runs if a goroutine, with errors reported through gErrors.
+func getFromCluster(f func(params *content.Params) (string, error), params *content.Params, fileName string, wg *sync.WaitGroup) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		out, err := f(params)
+		appendGlobalErr(err)
```

Comment:
better to check nil before append to global otherwise even for nil error case we have to lock the slice?

Reply:
done, in the helper func.